### PR TITLE
fix: Email COA's from Sales Invoice

### PR DESCRIPTION
**Pull request by hitenshKharva**
Ref: [Asana Task](https://app.asana.com/0/1199082161983365/1200166130781170)

image

### CODE REVIEW

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 1]
```diff
@@ -1590,3 +1590,33 @@ def create_invoice_discounting(source_name, target_doc=None):
 def get_item_details(item_code):
 	item = frappe.get_doc("Item", item_code).as_dict()
 	return item
+
+@frappe.whitelist()
+def email_coas(docname):
+	sales_invoice = frappe.get_doc("Sales Invoice", docname).as_dict()
+
+	for item in sales_invoice.get("items"):
+		delivery_note = frappe.get_doc("Delivery Note", item.delivery_note).as_dict()
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
		delivery_note = frappe.db.get_value("Delivery Note", item., "contact_email"delivery_note)
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 2]
```diff
@@ -436,6 +436,23 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
 		this.frm.refresh_field("outstanding_amount");
 		this.frm.refresh_field("paid_amount");
 		this.frm.refresh_field("base_paid_amount");
+	},
+
+	email_coas: function(doc) {
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]:
```suggestion
	email_coa: function(frm) {
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 3]
```diff
@@ -436,6 +436,23 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
...
+		frappe.call({
+			method: "erpnext.accounts.doctype.sales_invoice.sales_invoice.email_coas",
+			args: {
+				docname: doc.name
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]:
```suggestion
				docname: frm.doc.name
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 4]
```diff
@@ -436,6 +436,23 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
...
+			},
+			callback: (r) => {
+				if(r.message == "success"){
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]:
```suggestion
				if(r && r.message == "success"){
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 5]
```diff
@@ -436,6 +436,23 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
...
+					frappe.show_alert({
+						indicator: 'green',
+						message: __('Email Queued')
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]:
```suggestion
						message: __('Email Sent')
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.json [hunk 6]
```diff
@@ -62,6 +62,7 @@
   "update_stock",
   "scan_barcode",
   "items",
+  "email_coas",
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.json]:
```suggestion
  "email_coa",
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.json [hunk 7]
```diff
@@ -1604,12 +1605,18 @@
    "fieldtype": "Link",
    "label": "Coupon Code",
    "options": "Coupon Code"
+  },
+  {
+   "depends_on": "eval:doc.docstatus==1",
+   "fieldname": "email_coas",
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.json]:
```suggestion
   "fieldname": "email_coa",
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 8]
```diff
@@ -1590,3 +1590,33 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+
+		if not delivery_note.get("contact_email"):
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
		if not delivery_note:
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 9]
```diff
@@ -1590,3 +1590,33 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+			frappe.throw(_("No contact email found"))
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
			frappe.throw(_("No contact email found in Delivery Note {0}").format(item.delivery_note))
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 10]
```diff
@@ -1590,3 +1590,33 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+
+		coas = item.get("certificate_of_analysis")
+
+		if not coas:
+			frappe.msgprint(_("No Certificates of Analysis attached"))
+			return
+
+		attachments = []
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
		attachments = []
```

- Only attachments for last item will be sent.
- move attachments list out of for loop

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 11]
```diff
@@ -1590,3 +1590,30 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+	attachments = []
+	for item in sales_invoice.get("items"):
+		delivery_note = frappe.db.get_value("Delivery Note", item.delivery_note, "contact_email")
+		if not delivery_note:
+			frappe.throw(_("No contact email found in Delivery Note {0}").format(item.delivery_note))
+
+		coas = frappe.db.get_value("Delivery Note Item", {'item_code': item.item_code}, 'certificate_of_analysis')
+
+		if not coas:
+			frappe.msgprint(_("No Certificates of Analysis attached"))
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
			frappe.msgprint(_("No Certificates of Analysis attached for Item {0} in Delivery Note {1}.").format(item.item_code, item.delivery_note))
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 12]
```diff
@@ -436,6 +436,23 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
 		this.frm.refresh_field("outstanding_amount");
 		this.frm.refresh_field("paid_amount");
 		this.frm.refresh_field("base_paid_amount");
+	},
+
+	email_coa: function(doc) {
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]:
```suggestion
	email_coa: function() {
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 13]
```diff
@@ -436,6 +436,23 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
...
+		frappe.call({
+			method: "erpnext.accounts.doctype.sales_invoice.sales_invoice.email_coa",
+			args: {
+				docname: doc.name
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]:
```suggestion
				docname: this.frm.doc.name
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 14]
```diff
@@ -1590,3 +1590,30 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+			return
+
+		coa_file_id = frappe.db.get_value("File", {"file_url": coas}, "name")
+		attachments.append({"fid": coa_file_id})
```
Comment (sahil28297) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1626]:
```suggestion

		if not coa_file_id:
			frappe.msgprint(_("No File found for {0}").format(coas))
			return

		attachments.append({"fid": coa_file_id})
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 15]
```diff
@@ -1590,3 +1590,30 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+
+@frappe.whitelist()
+def email_coa(docname):
+	sales_invoice = frappe.get_doc("Sales Invoice", docname).as_dict()
+
+	attachments = []
+	for item in sales_invoice.get("items"):
+		delivery_note = frappe.db.get_value("Delivery Note", item.delivery_note, "contact_email")
+		if not delivery_note:
+			frappe.throw(_("No contact email found in Delivery Note {0}").format(item.delivery_note))
```
Comment (sahil28297) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
			frappe.msgprint(_("No contact email found in Delivery Note {0}").format(item.delivery_note))
			return
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 16]
```diff
@@ -1590,3 +1590,36 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+@frappe.whitelist()
+def email_coa(docname):
+	sales_invoice = frappe.get_doc("Sales Invoice", docname).as_dict()
+
+	attachments = []
+	for item in sales_invoice.get("items"):
+		delivery_note = frappe.db.get_value("Delivery Note", item.delivery_note, "contact_email")
+		if not delivery_note:
+			frappe.msgprint(_("No contact email found in Delivery Note {0}").format(item.delivery_note))
+			return
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
			continue
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 17]
```diff
@@ -1590,3 +1590,36 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+
+		coas = frappe.db.get_value("Delivery Note Item", {'item_code': item.item_code}, 'certificate_of_analysis')
+
+		if not coas:
+			frappe.msgprint(_("No Certificates of Analysis attached for Item {0} in Delivery Note {1}.").format(item.item_code, item.delivery_note))
+			return
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
			continue
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 18]
```diff
@@ -1590,3 +1590,36 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+
+		coa_file_id = frappe.db.get_value("File", {"file_url": coas}, "name")
+
+		if not coa_file_id:
+			frappe.msgprint(_("No File found for {0}").format(coas))
+			return
```
Comment (hrwX) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
			continue
```

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — DISMISSED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 19]
```diff
@@ -1590,3 +1590,36 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+			frappe.msgprint(_("No File found for {0}").format(coas))
+			continue
+
+		attachments.append({"fid": coa_file_id})
+
+	frappe.sendmail(
+		recipients=delivery_note.get("contact_email"),
+		subject="Certificate of Analysis",
+		attachments=attachments
+	)
```
Comment (sahil28297) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]:
```suggestion
	if delivery_note.get("contact_email") and attachments:
		frappe.sendmail(
			recipients=delivery_note.get("contact_email"),
			subject="Certificate of Analysis",
			attachments=attachments
		)
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 20]
```diff
@@ -1590,3 +1590,37 @@ def create_invoice_discounting(source_name, target_doc=None):
 def get_item_details(item_code):
 	item = frappe.get_doc("Item", item_code).as_dict()
 	return item
+
+@frappe.whitelist()
+def email_coa(docname):
```
Comment (rvpasari) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1595]: Where is the docstring?

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 21]
```diff
@@ -1590,3 +1590,37 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+	sales_invoice = frappe.get_doc("Sales Invoice", docname)
+	attachments = []
+
+	for item in sales_invoice.items:
```
Comment (rvpasari) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1607]: comments for each block

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 22]
```diff
@@ -1590,3 +1590,37 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+		delivery_note = frappe.db.get_value("Delivery Note", item.delivery_note, "contact_email")
+		if not delivery_note:
+			frappe.msgprint(_("No contact email found in Delivery Note {0}").format(item.delivery_note))
```
Comment (rvpasari) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1611]: Have you tested if this message print works? Does the .format() have to be inside the _("") or outside?

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 23]
```diff
@@ -1590,3 +1590,37 @@ def create_invoice_discounting(source_name, target_doc=None):
...
+
+		coa_file_id = frappe.db.get_value("File", {"file_url": coas}, "name")
+
+		if not coa_file_id:
+			frappe.msgprint(_("No File found for {0}").format(coas))
+			continue
+
+		attachments.append({"fid": coa_file_id})
+
+	if delivery_note.get("contact_email") and attachments:
```
Comment (rvpasari) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py]: add description in comments

**Review by rvpasari — CHANGES_REQUESTED** [review]

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py (hunk 22 above)
Comment (hitenshKharva) [reply erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1602]: image

yes

Comment (hitenshKharva) [reply erpnext/accounts/doctype/sales_invoice/sales_invoice.py:1611]: Yes. Please find the attached screenshot for your reference.

image

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**